grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +60,546 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
+
+      if descriptor:
+        self._ExtractSchema(descriptor, visiting)
+
+      schema_or_ref_obj = \
+        self._GetSchemaOrReferenceObject(
+          self._GetTypeName(field_descriptor),
+          field_descriptor.label == protobuf2.LABEL_REPEATED)
+
+      schema_obj["properties"][field_name] = schema_or_ref_obj
+
+    visiting.remove(t_name)
+    self.schema_objs[t_name] = schema_obj
+
+    return schema_obj
+
+  def _ExtractSchema(self, t, visiting: Set[str]):
+    if t is None:
+      raise ValueError(f"Trying to extract schema of None.")
+
+    t_name = self._GetTypeName(t)
+    # "Primitive" types should be already present in self.schema_objs.
+    if t_name in self.schema_objs:
+      return self.schema_objs[t_name]
+
+    if t_name in visiting:
+      # Dependency cycle.
+      return
+
+    if isinstance(t, Descriptor):
+      return self._ExtractMessageSchema(t, visiting)
+
+    if isinstance(t, EnumDescriptor):
+      return self._ExtractEnumSchema(t, visiting)
+
+    raise TypeError(f"Don't know how to handle type \"{t_name}\" "
+                    f"which is not a protobuf message Descriptor, "
+                    f"nor an EnumDescriptor, nor a primitive type.")
+
+  def _ExtractSchemas(self):
+    """Extracts OpenAPI schemas for all the used protobuf types."""
+
+    self.schema_objs = dict()  # Holds OpenAPI representations of types.
+    self._AddPrimitiveTypesSchemas()
+
+    visiting = set()  # Holds state of types extraction (white/gray nodes).
+    router_methods = self.router.__class__.GetAnnotatedMethods()
+    for method_metadata in router_methods.values():
+      args_type = method_metadata.args_type
+      if args_type:
+        if inspect.isclass(args_type) \
+            and issubclass(args_type, rdf_structs.RDFProtoStruct):
+          self._ExtractSchema(args_type.protobuf.DESCRIPTOR, visiting)
+        else:
+          self._ExtractSchema(args_type, visiting)
+
+      result_type = method_metadata.result_type
+      if result_type:
+        if inspect.isclass(result_type) \
+            and issubclass(result_type, rdf_structs.RDFProtoStruct):
+          self._ExtractSchema(result_type.protobuf.DESCRIPTOR, visiting)
+        else:
+          self._ExtractSchema(result_type, visiting)
+
+  def _SetComponents(self):
+    if self.schema_objs is None:
+      raise ValueError("Called _SetComponents before extracting schemas.")
+
+    # The Components Object "components" of the root OpenAPI object.
+    components_obj = dict()
+    schemas_obj = dict()
+
+    type_names = set(self.schema_objs.keys())
+    primitive_types_names = set(self.primitive_types_names)
+    # Create components only for composite types.
+    for type_name in type_names - primitive_types_names:
+      schemas_obj[type_name] = self.schema_objs[type_name]
+
+    components_obj["schemas"] = schemas_obj
+
+    self.open_api_obj["components"] = components_obj
+
+  def _GetSchemaOrReferenceObject(self, type_name: str, is_array=False):
```

Comment:
Missing type annotation (here and in the all methods below).

Reply:
Added them. Thank you!